### ISSUE TITLE
TestScheduler is unusable

### ISSUE THREAD

**Issue by ghostbuster91** [monix/monix#1095]
Hi, I don't know if I am using testScheduler wrong or it is something fundamentally wrong with it.

Consider following simple code:
```
val ts = TestScheduler()

    val tick = Observable
      .interval(1.second)
      .mapEval(_ => Task.eval(println("executing")))
      .mapEval(_ => Task.sleep(1 seconds).executeOn(Scheduler.global))
      .mapEval(_ => Task.eval(println("after sleep")))

    tick.subscribe()(ts)
    Thread.sleep(5000)
```
1. Why this observable is being executed in the first place without invoking `ts.tickOne()`? How am I supposed to launch my observable in a controllable fashion? Keep in mind that this is a simple example and in the real life I might not be in control of calling `subscribe`.

2. I was expecting that after calling `ts.tickOne()` I would eventually see `"after sleep"` but this is not the case since this code expects me to call `ts.tickOne` again but only after `Task.sleep(1 seconds).executeOn(Scheduler.global)` finishes. So in other words I have to put `Thread.sleep(x)` in my test to wait for this action to finish. If so what's the purpose of using testScheduler in the first place? Again this is a simple example and in the real life I might not even know how long this action will take. Forcing the observable to execute two or more times is equally hard.

I can provide an example which is more connected to the real life if you would like.

Update: I have checked that similar code written using rxJava works as expected:
```
    val ts = new TestScheduler()
    val tick = Observable
      .interval(0, 1, TimeUnit.SECONDS, ts)
      .flatMapSingle(
        _ =>
          Single.fromCallable(new Callable[Unit] {
            override def call(): Unit = {
              println("before sleep")
            }
          })
      )
      .flatMapSingle(_ => Single.just(1).delay(1, TimeUnit.SECONDS, Schedulers.io()))
      .flatMapSingle(
        _ =>
          Single.fromCallable(new Callable[Unit] {
            override def call(): Unit = {
              println("after sleep")
            }
          })
      )

    tick.subscribe()

    ts.triggerActions()
    Thread.sleep(5000)
```

**Comment by Avasil** [comment]
> Why this observable is being executed in the first place without invoking ts.tickOne()? How am I supposed to launch my observable in a controllable fashion? Keep in mind that this is a simple example and in the real life I might not be in control of calling subscribe.

Running `Observable` does not introduce any async boundaries so your test most likely executes on `main` or `global` thread pool. `Observable.interval` runs the first tick synchronously without touching scheduler. I guess `rxJava` implementation always schedule the next tick asynchronously. It might be good idea to do it as well, I need to think about it. The best fix right now is to call `tick.executeAsync.subscribe()(ts)`.

> I was expecting that after calling ts.tickOne() I would eventually see "after sleep" but this is not the case since this code expects me to call ts.tickOne again but only after Task.sleep(1 seconds).executeOn(Scheduler.global) finishes.

Consider how `tickOne` works:

```
* Executes just one tick, one task, from the internal queue, useful
* for testing that some runnable will definitely be executed next.
```

What's in the internal queue?

```scala
  tick.executeAsync.subscribe()(ts)

  // TreeSet(Task(1,monix.reactive.internal.operators.SubscribeOnObservable$$anon$1@69fb6037,0 days))
  println(ts.state.tasks)
  // executing
  ts.tickOne()
  // TreeSet()
  println(ts.state.tasks)
```
There is only one task. Once we execute it, we get to`mapEval(_ => Task.sleep(1 seconds).executeOn(Scheduler.global))` but it uses different `Scheduler` so we can't progress it.

When it's done, we get a new task so we have something that we can `tick()` again:

```
  Thread.sleep(1500)
  // TreeSet(Task(2,<function1>,0 days))
  println(ts.state.tasks)
```

If we tick one by one we will see that it is going quite slowly because there is a couple of tasks created:

```
  // after sleep
  ts.tickOne()
  // TreeSet(Task(3,Future(<not completed>),0 days))
  println(ts.state.tasks)
  ts.tickOne()
  // TreeSet(Task(4,Future(<not completed>),0 days))
  println(ts.state.tasks)
  ts.tickOne()
  // TreeSet(Task(5,monix.reactive.internal.builders.IntervalFixedDelayObservable$$anon$1@f78a47e,1 second))
  println(ts.state.tasks)
```

What about `tick()`?

```
    * Triggers execution by going through the queue of scheduled tasks and
    * executing them all, until no tasks remain in the queue to execute.
```

So if we use it instead, we will go through everything that doesn't involve timer:

```
// after sleep
ts.tick()
// TreeSet(Task(5,monix.reactive.internal.builders.IntervalFixedDelayObservable$$anon$1@644baf4a,1 second))
println(ts.state.tasks)
```

We can also try `ts.tick(100.days)` instead and the result will be as follows:

```
// after sleep
// executing
ts.tick(100.days)
// TreeSet()
println(ts.state.tasks)
```

Once again, it gets to the point of execution on different `Scheduler`  and doesn't have any tasks left to do. There is no way to know that after executing that task there will be new ones. 

The only way I can think of to "run all tasks that don't require any sleep" would be busy polling in the background and I'm not sure if it fits well with `TestScheduler` as it is. If there is a need, the users can always do it themselves or have something like 

```scala
while( ts.state.tasks.isEmpty) {}
```

instead of `Thread.sleep`. Well, preferably something with a timeout like `eventually` in ScalaTest.

> If so what's the purpose of using testScheduler in the first place?

The main use case is for executing stuff on `TestScheduler`s only, I recommend to take a look at the many tests for `Task`, `Observable` or even in `Cats-Effect` for examples. 

Interacting with other thread pools is tricky but doable

> Forcing the observable to execute two or more times is equally hard.

```
  // executing
  ts.tickOne()
  while (ts.state.tasks.isEmpty) {}
  // after sleep
  // executing
  ts.tick(1.second)
  while (ts.state.tasks.isEmpty) {}
  // after sleep
  // executing
  ts.tick(1.second)
```

as a function:

```scala
  // executing
  ts.tickOne()

  executeN(10)

  def executeN(times: Int): Unit = (0 until times) foreach { _ =>
    while (ts.state.tasks.isEmpty) {}
    // after sleep
    // executing
    ts.tick(1.second)
  }
```

**Comment by ghostbuster91** [comment]
Thank you very much for the detailed response. 
> I guess rxJava implementation always schedule the next tick asynchronously. It might be good idea to do it as well, I need to think about it. 

Let me know what you came up with, maybe this will be a good idea for my first contribution to monix ;)

In terms of `Scheduler` I will probably try to write my own which does the busy pooling as you suggested.

**Comment by Avasil** [comment]
I've created an issue: https://github.com/monix/monix/issues/1185
